psycopg/psycopg/copy.py

```diff
@@ -629,14 +645,74 @@ async def write(self, data: Buffer) -> None:
             for i in range(0, len(data), MAX_BUFFER_SIZE):
                 await self._queue.put(data[i : i + MAX_BUFFER_SIZE])
 
-    async def finish(self, exc: Optional[BaseException] = None) -> None:
+    async def finish(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> None:
         await self._queue.put(b"")
 
         if self._worker:
             await asyncio.gather(self._worker)
             self._worker = None  # break reference loops if any
 
-        await super().finish(exc)
+        await super().finish(exc_type, exc_val, exc_tb)
+
+
+try:
+    import anyio
+    import anyio.abc
+except ImportError:
+    pass
+else:
+
+    class AnyIOLibpqWriter(AsyncLibpqWriter):
```

Comment:
In 3.1, this seems more the behaviour of `AsyncQueuedLibpqWriter` than `AsyncLibpqWriter`, because it spawn separate workers, as opposite as `AsyncLibpqWriter`, which works in-task.

Afaics, `AsyncLibpqWriter` would work out-of-the-box with anyio, because it only uses `await self.connection.wait(<some generator>)` to send control back to the loop.

Originally, using a separate thread/task seemed to yield some results, but in following tests I couldn't verify a real benefit for using `AsyncQueuedLibpqWriter`. As such, psycopg 3.1 uses `AsyncLibpqWriter` by default, and maybe it's not worth to have an `AnyIOLibpqWriter`.

Reply:
Yes, I noticed that lately. I agree we should use `AsyncLibpqWriter` by default with anyio and `AnyIOLibpqWriter` (possibly renamed) would be the optional counterpart of `AsyncQueuedLibpqWriter`.